### ISSUE TITLE
RPA server needs to have normal wallet concepts to be complete

### ISSUE THREAD

**Issue by imaginaryusername** [Electron-Cash/Electron-Cash#2805]
Currently RPA has two known paths of privacy leaks in implementation:

1. Server leaks: RPA still subscribes to the "correct" addresses immediately, leaking the information unnecessarily to the Fulcrum/Rostrum server it's connected to. This is fixable trivially by running your own server, so while annoying for many isn't fatal, and is fixable soon ish? 

2. Wallet behavior: Right now an RPA wallet is basically an imported-address wallet that imports from server, which has **major** consequences for privacy in normal use. When spending anything, it sends change back to the spending address: Not only does this leak which address is the "right" address, it also makes it inevitable that eventually two coins would be spent together creating linkage that RPA was designed to avoid in the first place.

To solve #2 it's likely that RPA wallets will somehow need:

1. Change addresses, perhaps derived from seed in HD; and

2. Enabling of Cashfusion to safely absorb incoming coins without linking them

The task is probably very tedious as it involves heavy lifting on the RPA wallet structure, but it can be done, and the work is essential. :pray:

**Comment by fyookball** [comment]
For now the intended use is that you would forward your incoming RPA funds to a Standard wallet type.  

Change addresses are non trivial here because the keystore operates more like imported wallet than HD.  Perhaps a good approach might be to link an RPA wallet to another standard wallet and have some kind of automated transfer.

**Comment by minisat0shi** [comment]
Automated transfer would be great

**Comment by imaginaryusername** [comment]
Yes, of course it's nontrivial, but in practical use we can't exactly expect users to manually go to the coins tab and send each incoming coins 1/1 to another wallet (which creates a pattern of 1/1 sends that's problematic itself, so presumably you even need to somehow fake a change trail while doing that). It's not sustainable for public consumption. 

There may be creative ways to piggyback an HD wallet in there somehow, but I don't know the details.

**Comment by cculianu** [comment]
There’s a lot of engineering to do and it will still be imperfect.

RPA as a concept is very good and is very private potentially.

But the entire design of Electrum protocol and electron cash is going to leak info to servers.

we can over complicate it and engineer some very imperfect complicated solution and it will still leak privacy to servers

I really don’t think this should be the focus of electron cash right now, personally… especially since it would cost resources to do this and lead to more and more code to maintain… and still not be perfect.

One can run a neutrino style wallet that talks p2p directly… and implement that separately. That’s probably gunna be more private.

I’m not sure we have the resources or the interest to solve this really fully.

As for fusion: RPA wallet users can just send funds to their HD wallets if they want to fuse.

I’d categorize this as a nice list of issues but not prioritize this personally.

**Comment by cculianu** [comment]
> For now the intended use is that you would forward your incoming RPA funds to a Standard wallet type.
> 
> Change addresses are non trivial here because the keystore operates more like imported wallet than HD. Perhaps a good approach might be to link an RPA wallet to another standard wallet and have some kind of automated transfer.

Yes this

**Comment by imaginaryusername** [comment]
As said, leaking to server isn't even the biggest issue here... leaking on the blockchain however isn't a thing solved by trust. It's kind of a big deal.

Also as said elsewhere, I'm not saying this needs to be solved today or tomorrow, yes there are other pressing issues and I understand. But since we got RPA in here, we should understand what would it take to fulfill its promise at least in some practical way. 

RPA is a solution that tackles multiple things btw, chief among them allowing for a single identity that Neutrino absolutely doesn't deal with - it only deal with server information leakage, it doesn't deal with any information problems on the blockchain. So I don't think the comparison to Neutrino is quite apt here.

**Comment by cculianu** [comment]
I thought about it -- and this is how it should work. This needs a redesign.

1. CashFusion should be required.
2. All all *all* HD standard wallets (that can fuse) should have this RPA Paycode thing as an optional feature. If enabled, then the RpaManager kicks off and scans the chain in the background. 
3. This means all wallets that can fuse get an optional PayCode "for free"
4. RpaManager, if it finds a utxo and discoveres a new privkey .. either hands it off to fusion  (needs work in CashFusion to support) -- or for now -- Sends to the normal HD wallet addresses for fusing.
5. RpaManager refuses to do anything unless Fusion is running and "spend only fused coins" is enabled...
6. No new wallet type is/was needed.  WE HAVE TOO MANY WALLET TYPES ALREADY.  A standard wallet could have been used, with the extra optional "RpaManager" for discovering utxos/privkeys.  So you can opt-in for this feature instantly with your existing normal wallet...
7. No spending of a discovered paycode utxo possible until it's been "laundered" through Fusion.

The above design would be SIMPLER than the existing one and would have been SELF CONTAINED within a single "RpaManager" that has to worry about discovering UTXOs from the paycode and handing them off to fuse.

I would not create any significant technical debt as compared to other solutions -- such as YET ANOTHER wallet type with MULTIPLE keystores (no no no no NO! Please NO!).

The design proposed above is the best design for preserving privacy... and .. bonus -- **every wallet now has an optional paycode**. And double bonus: all changes (mostly) self-contained within a manager class -- RpaManager -- and no crazy spaghetti changes everywhere and NO ADDITIONAL WALLET TYPES NEEDED.

-Calin

**Comment by fyookball** [comment]
@cculianu sounds good overall but I don't understand why we would not need the second keystore.  The RPA transactions are (practically speaking) unrelated to the HD keystore.

**Comment by cculianu** [comment]
@fyookball I would prefer it if we continue implementation detail level discussions about data structures (a keystore is just a tiny detail) -- if we keep those to when we are implementing and we do that in telegram or phone calls, rather than here. It feels distracting to talk about this here rather than the big picture.
